fix(platform-browser): unencapsulated styles leaking out of shadow dom host

crisbeto opened the pull request:
Currently when a component with `ViewEncapsulation.None` is inserted, it always adds its styles to `document.head`. This can be problematic when the component is inside the shadow DOM, because it can cause styles to leak out of it.

These changes resolve the issue by adding some extra logic to the `SharedStylesHost` that allows for multiple "insertion nodes" to be associated with a component host node. Then we use this new functionality to associate the unencapsulated component styles with the closest shadow root.

Note that in most cases we'll have one insertion node per host node, but we have to support multiple for the case where several unencapsulated components are inserted inside a shadow DOM host.

Fixes #35039.

JoostK commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -114,6 +114,14 @@ export class DomRendererFactory2 implements RendererFactory2 {
+        // styles leak out into the document.
+        const rootNode = element.getRootNode?.();
+        if (typeof ShadowRoot !== 'undefined' && rootNode instanceof ShadowRoot) {
+          return new ShadowDomRenderer(
```
I noticed that `ShadowDomRenderer` is doing style flattening in its constructor for every instantiation, instead of doing it only once and caching the result somewhere. This includes the regex replace for the `%COMP%` pattern, but that won't even be present as that is only applied by the compiler for components that use emulated style encapsulation. We can probably optimize this a bit, which has become more relevant now that the shadow DOM renderer is used more often (also because this can now trigger usage of `ShadowDomRenderer` even if not used in the Angular app itself, but bootstrapped into a shadow tree).

JoostK commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -307,14 +315,16 @@ class EmulatedEncapsulationDomRenderer2 extends DefaultDomRenderer2 {
-    this.sharedStylesHost.addHost(this.shadowRoot);
+    this.shadowRoot = ancestorShadowRoot || hostEl.attachShadow({mode: 'open'});
+    this.ownsShadowRoot = !ancestorShadowRoot;
+    this.sharedStylesHost.addHost(hostEl, this.shadowRoot);
```
Wouldn't this mean that multiple non-encapsulated host nodes for the same component add their styles into the shadow root multiple times?

crisbeto replied:
I hadn't really considered that, but I think you're right. I suppose that we can track the inserted styles on `SharedStylesHost._hostNodes`, but that'll complicate the data structure even more.

crisbeto commented on packages/platform-browser/src/dom/shared_styles_host.ts:
```diff
@@ -34,44 +34,61 @@ export class SharedStylesHost {
-  // Maps all registered host nodes to a list of style nodes that have been added to the host node.
-  private _hostNodes = new Map<Node, Node[]>();
+  // Mapping of component host nodes to the nodes in which their styles are inserted.
+  private _hostNodes = new Map<Node, Map<Node, Node[]>>();
```
There can be multiple insertion nodes per host node in the following case:

```html
<uses-shadow-dom>
  <unencapsulated-comp></unencapsulated-comp>
  <other-unencapsulated-comp></other-unencapsulated-comp>
</uses-shadow-dom>
```

In this case the hosts are `unencapsulated-comp` and `other-unencapsulated-comp` while the insertion node is the shadow root of `uses-shadow-dom`. We don't want to clear all styles associated with `uses-shadow-dom` when one of the unencapsulated components is destroyed.

crisbeto replied:
There is, but how would we distinguish which styles belong to which host node? I was trying to solve the issue where multiple host nodes point to a single insertion node and then one of the hosts is removed. In this case we don't want to clear the entire insertion node.

JoostK replied:
Would this be sufficient:

```ts
private _hostNodes = new Map<Node, { insertionPoint: Node; styles: Node[] }>();
```

Then removing a host node can remove the corresponding styles from its insertion point. But I realize that the above is still insufficient when there's multiple host nodes that share styles that they have attached into the same insertion point; that is a bit more hard to track.

crisbeto replied:
So I think that we could do it by changing `SharedStylesHost._stylesSet` to a `Map<string, Set<Node>>` where the keys are the inserted styles and the set is the insertion nodes that have received the styles. My concern is that this will slow down cleanup, because we'd have to look through all the styles in order to remove the insertion node reference.

crisbeto replied:
I pushed a change that should simplify this and resolve the issue where the same styles are inserted multiple times. It uses the insertion node as the key and keeps a counter of host nodes associated with the insertion node. The only downside is that we'll only clear the styles once every host node has been destroyed, but that doesn't seem like it would be too much of an issue?

crisbeto committed:
fix(platform-browser): unencapsulated styles leaking out of shadow dom host

Currently when a component with `ViewEncapsulation.None` is inserted, it always adds its styles to `document.head`. This can be problematic when the component is inside the shadow DOM, because it can cause styles to leak out of it.

These changes resolve the issue by adding some extra logic to the `SharedStylesHost` that allows for multiple "insertion nodes" to be associated with a component host node. Then we use this new functionality to associate the unencapsulated component styles with the closest shadow root.

Note that in most cases we'll have one insertion node per host node, but we have to support multiple for the case where several unencapsulated components are inserted inside a shadow DOM host.

Fixes #35039.

crisbeto committed:
fixup! fix(platform-browser): unencapsulated styles leaking out of shadow dom host

jessicajaniuk commented:
@crisbeto Did you still want to land this PR?

crisbeto commented:
Yes, I think that all the comments were addressed but the discussion has stalled.

jesusreal commented:
Also waiting for this to be merged. It would be great to have this soon in Angular. Are there any news on it?

mangei commented:
Hi, any news here. We ran into the same issue, when providing our Angular webcomponents which led to break the style of the host-application. (both applications used Angular Material, but in different versions)

robert-isaac-sap commented:
shouldn't this be closed because of https://github.com/angular/angular/pull/58482
